[SYCL][Doc] Cluster Group Extension Document

AD2605 opened the pull request:
Initial public working draft for thread block cluster support in SYCL, intended to get feedback.

Contains the proposal for - 
1. Launching a kernel with cluster group
2. Accessing the various `ids` associated with the cluster_group from the kernel
3. Cluster level barrier
4. Accessing another workgroup's local memory

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+
+== Overview
+Nvidia's compute capability (cc) 9.0 devices introduced a new level in the thread hierarchy, called as thread block clusters, in CUDA terminology. A thread block cluster, is a collection of thread blocks (a workgroup in SYCL terminology), which are guaranteed to be scheduled to be scheduled on a separate compute unit, and are guaranteed to run in parallel, and the thread blocks in the cluster can ...
```
Add copyright notices to all mentions of NVIDIA and a general disclaimer at the right section

AD2605 replied:
Done

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+
+== Overview
+Nvidia's compute capability (cc) 9.0 devices introduced a new level in the thread hierarchy, called as thread block clusters, in CUDA terminology. A thread block cluster, is a collection of thread blocks (a workgroup in SYCL terminology), which are guaranteed to be scheduled to be scheduled on a separate compute unit, and are guaranteed to run in parallel, and the thread blocks in the cluster can ...
```
```suggestion
Nvidia's compute capability (cc) 9.0 devices introduced a new level in the thread hierarchy, called as thread block clusters, in CUDA terminology. A thread block cluster, is a collection of thread blocks (a workgroup in SYCL terminology), which are guaranteed to be scheduled on a distinct compute unit, and are guaranteed to run in parallel, and the thread blocks in the cluster can access each other's shared memory (local memory SYCL terminology). This has various applications, convolutions, GEMMs and FFTs to name a few.
```

Ruyk replied:
Could you add a link to the NVIDIA documentation of the feature?

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+=== Launching a kernel with work group clusters
+
+To launch a kernel with a cluster dimension, because of the special scheduling guarantees, the driver requires which kernel will be using the said feature. This introduces an additional parallel_for overload and parallel_for_work_group overloads, accepting cluster dimensions as a `sycl::range`, the dimension of which will be the same as that of the global and the local range.
```
I don't think we should not introduce this for parallel for work group

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+== Status
+
+This is a proposed extension specification, intended to gather community
```
Is this a general SYCL extension you are proposing? or an NVIDIA backend specific one?

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+=== Accessing the Work Group Cluster from within the kernel
+
+Building upon the group heirarchy in SYCL, this proposal another heirachy in the group after `sub_group` and `group` (for workgroups), a class to be called as `cluster_group`, which will  be accessible via the `nd_item` class, via a method to be introduced called `get_cluster_group()`
```
```suggestion
Building upon the group hierarchy in SYCL, this proposal another hierarchy in the group after `sub_group` and `group` (for workgroups), a class to be called as `cluster_group`, which will  be accessible via the `nd_item` class, via a method to be introduced called `get_cluster_group()`
```

run ispell  on the text :-)

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+|`size_t get_local_linear_range(void)`
+|Returns the total number of `workItems` within the cluster
+|===
```
all the above methods should be const right?

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|===
+
+
+CC 9.0 devices feature the ability to synchronize all the workgroups within the cluster, and to achieve that in SYCL, this proposes add an overload to the existing `sycl::group_barrier` function, accepting `cluster_group` type.
```
Explain CC acronym somewhere

AD2605 replied:
Done, added a glossary section

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+To prevent un-defined behaviour, in the case where these range's or id's are queried without the cluster dimensions mentioned at launch can be considered as a cluster launch with cluster dimension as 1.
+
+=== Accessing another workgroup's local memory
```
I suggest we move this to another extension. That way we can get them approved/implemented independently.

Ruyk reviewed (changes requested):
Thanks for the proposal, please also add an example (or two) of how it would be used

jbrodman commented:
Note: This direction is not approved. Do not merge this as the APIs and abstractions are not final.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+=== Accessing another workgroup's local memory
+
+Workgroup's within the cluster have the ability to access another workgroup's local memory. The local memory addresses can only be addressed by the `workItems` of that workgroup. Therefore, to access another workgroup's local memory, the address would need to be mapped such that it is addressable by calling `workItem` belonging to another workgroup. A public member function of the `cluster_group` ...
```
```suggestion
Work-group's within the cluster have the ability to access another work-group's local memory. Typically addresses which reside in the local memory of a work-group can only be addressed by the work-items of that work-group. Therefore, to access another work-group's local memory, the address needs to be mapped such the address in another work-group is addressable within the calling work-item. A member function of the `cluster_group` class; `get_cluster_local_pointer` will perform the above and return a pointer which can then be dereferenced by the calling work-item.
```

AerialMantis replied:
I think we have this now in `map_cluster_local_pointer`, I'll re-review the latest revision.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+[%header,cols="10,5"]
+|===
+
+|`T*  get_cluster_local_pointer(T* addr, size_t group_id)`
```
I would maybe suggest naming this something like `map_cluster_local_pointer` instead since it's doing more than just getting a pointer.

AerialMantis replied:
I think it would make sense for this interface to use `multi_ptr` to constrain it to pointers to local memory, and avoid UB. @Naghasan what do think think?

AerialMantis replied:
What are the control-flow semantics of the equivalant feature in CUDA, does this need to be called in convergent control flow, and must all work-items in the work-group take the same parameters?

Also is this function synchronous, when it returns is the pointer returned immediately accessible in the current work-group?

AerialMantis replied:
> Each thread can call this independently. The only requirement is that the thread block of which the shared memory is being accessed, should have started executing and not finished execution, which can be guaranteed by a cluster wide synchronization. (source Cuda programming guide)

So typically group-based functions in SYCL are convergent, in that all work-items in a particular scope must call the function in the same control flow, however, I believe in order to use this feature effectively it must be called per-work-item, and will only work on hardware with independent forward progress of work-items.

In the future if we come to generalise this extension across all backends I think we will need to address this, but since this is CUDA-specific I think it is fine for now. Though we should explain these semantics in definition of the function.

> The scope is per-thread, so when this pointer is returned it is accessible to that thread only

We should explain this in the definition of the function as well so it's clear that the mapping is blocking and will be complete by the time the function returns.

AD2605 replied:
> So typically group-based functions in SYCL are convergent, in that all work-items in a particular scope must call the function in the same control flow

But isn't that only for group algorithms which collaboratively compute / share data. I believe in this case there is no "collaboration" so to speak.

The only requirement to access another work-group's local memory, is that the work-group should exist, implying that local memory exists. Then Any work-item can map that local memory start address and start dereferencing it as required. 

> and will only work on hardware with independent forward progress of work-items.

Assume a device with the lowest level of independent forward progress at a sub-group level. On that device, even if only say 1 work-item calls this function, would this be conceptually different than - 
```cpp
if (sub_group.get_local_linear_id() == 0) {
    any_function_foo();
}
```
I say so because note, the work-item calling `map_cluster_local_pointer` does not **directly** depend on another work-item directly within it's group, as in any another work-item is not responsible to return the address, but rather the existence of another work-group within the cluster-group.

AD2605 replied:
> I say so because , the work-item calling map_cluster_local_pointer does not directly depend on another work-item within the group, as in any another work-item is not responsible to return the address, but rather the existence of another work-group within the cluster-group.

BTW, this is indeed very Nvidia specific, but it's a cuda only extension so it's fine for now. If another hardware starts to support this feature, and has another set of requirement, they would need to be incorporated somehow, depending on what they are.

But, one of my main discussions above is that maybe independent forward progress of work-items is not required for this feature from what I gather, seeing how it is exposed at the PTX level

AerialMantis replied:
>But isn't that only for group algorithms which collaboratively compute / share data. I believe in this case there is no "collaboration" so to speak.

That's right, this function I think we need to make. a clear distinction distinction between this kind of function and group functions/algorithms, even though these involve a group id.

> I say so because , the work-item calling this function does not directly depend on another work-item within the group, as in any another work-item is not responsible to return the address, but rather the existence of another work-group within the cluster-group.

So this is right, `map_cluster_local_pointer` can execute independently within the work-item with no dependency on other work-items in the group, however, depending on how the function lowers to instructions and whether the hardware supports independent forward progress of work-items, it may or may not be able to execute in separate work-items with different parameters concurrently.

But I think we can define it as a built-in function which executes within a work-item, which does not require convergence, permitting different parameters in each work-item, but leave it unspecified whether these are then executed concurrently or masked and executed in sequence. Though this is perhaps thinking too far in the future, as this is still a CUDA-specific extension.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|===
+
+|`T*  get_cluster_local_pointer(T* addr, size_t group_id)`
+|Accepts the equivalent address to the memory location relative to the calling `workItem` which is to be mapped from the local memory of the workgroup, as specified by the `group_id`, denoting the flattened `group_id` within the cluster
```
```suggestion
|Accepts the equivalent address to the memory location relative to the calling work-item which is to be mapped from the local memory of the work-group, as specified by the `group_id`, denoting the flattened `group_id` within the cluster.
```

AerialMantis replied:
Does this need to be linearized id, could this not just be the regular 'id'?

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+
+== Overview
+Nvidia's compute capability (cc) 9.0 devices introduced a new level in the thread hierarchy, called as thread block clusters, in CUDA terminology. A thread block cluster, is a collection of thread blocks (a workgroup in SYCL terminology), which are guaranteed to be scheduled to be scheduled on a separate compute unit, and are guaranteed to run in parallel, and the thread blocks in the cluster can ...
```
We typically keep to an 80 character line limit to make it easier for cmd line editors and limiting the scope of future changes.

AerialMantis replied:
No worries, thanks :)

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+This extension is written against the SYCL 2020 revision 8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
```
I think we should list the dependencies on other extension here, so if we're dependent on the enqueue functions extension and the compile-time properties extension, we can mention that here. @gmlueck is this the right place to put this, I'm not sure what the convention for that is?

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+=== Launching a kernel with work group clusters
+
+To launch a kernel with a cluster dimension, because of the special scheduling guarantees, the driver requires which kernel will be using the said feature. This introduces an additional parallel_for overload and parallel_for_work_group overloads, accepting cluster dimensions as a `sycl::range`, the dimension of which will be the same as that of the global and the local range.
```
I think it would also make sense to have an aspect for querying whether a device supports the cluster-group launch, though since this is an Nvidia-specific extension, this may not be strictly necessary atm. @gmlueck what are you thoughts on this?

AerialMantis replied:
> Using enqueue functions and properties now,
Though if @AerialMantis or maybe @gmlueck , if you could please once verify if simply defining a subscript operator would be okay to return the property values / or it's usage in general, it would be really helpful.

I think having a subscript operator on the property would be useful, and I think there is precident for this in other extensions, see the `work_group_size` property in [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc).

> I was planning on using sycl_ext_oneapi_device_architecture to check for CC 9.0, but just wanted to ask for other's opinion before committing to it. For now, I have not mentioned the behaviour of the backend in case this is used on unsupported devices, and can be implementation specific

I think it makes sense to check the compute capability as a way for the backend to verify that it can support a cluster group launch, though despite this being a CUDA-specific extension, I would reflect this as an aspect, and leave the architecture check as an implementation detail.

> On a similar topic, should we also extend get_info on what are the valid ranges for the cluster dimension ?

I think this is a good idea, it would be useful for users to be able to query what the maximum cluster group size can be, something like `info::device::max_cluster_group_size`. I wonder if it's possible for the maximum cluster group size to change based on the compilation of the kernel, in which case we may want to have a compiled kernel query as well, something like `info::kernel_device_specific::max_cluster_group_size`. In CUDA `cudaOccupancyMaxPotentialClusterSize` takes the kernel function as a parameter so that would suggest that this may be the case.

AD2605 replied:
Added the device query `ext_cuda_experimental_max_cluster_group_size`
following the naming convention for additional device queries.

AerialMantis reviewed (commented):
Thanks for putting it together, overall this looks really good, I've made a couple of design suggestions.

I would also suggest adding a synopsis which shows the whole `cluster_group` class, including namespaces, this helps to see everything together, and it would be good to have an example of how this would be used, this helps people understand how everything fits together.

AD2605 commented:
Hi all, thanks for the initial review, I have addressed the comments (except one regarding copyright notice, I am yet to add that) and this now ready for the re-review

CC: @jbrodman

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+
+
+== Overview
+Nvidia's compute capability (cc) 9.0 devices introduced a new level in the thread hierarchy, called as thread block clusters, in CUDA terminology. A thread block cluster, is a collection of thread blocks (a workgroup in SYCL terminology), which are guaranteed to be scheduled to be scheduled on a separate compute unit, and are guaranteed to run in parallel, and the thread blocks in the cluster can ...
```
So I think the description in the CUDA describes this behaviours, it doesn't use the term run concurrently, rather it says co-scheduled on the SM, but I think that should be equivalent - https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#thread-block-clusters.

We may also want to specify in this extension the forward progress guarantee properties of work-groups within a cluster group when launching a kernel with a cluster group range, based on the definitions introduced in https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc), though we'd need to confirm the forward progress guarantees of CUDA. Maybe this could be left as a known issue at the bottom of the extension for now.

AD2605 replied:
>  but I think that should be equivalent - https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#thread-block-clusters.

Yeah, it mentions that they are guaranteed to be co-scheduled  on the same GPC ( GPU Processing cluster).

> We may also want to specify in this extension the forward progress guarantee properties of work-groups within a cluster group

Thanks for raising this point. I did not cross my mind and this spec should definitely mention the forward progress guarantees, but I would like to look into it a bit further, and make an addendum later on. For now, as you suggested, I will be adding a known issue stating that forward progress guarantees are yet to be confirmed.

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,276 @@
+
+[source,c++]
+----
+    class cluster_group {
```
Templating on Dimensions now

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,276 @@
+    public:
+        using id_type = id<3>;
+        using range_type = range<3>;
+        using linear_id_type = uint32_t;
```
changed it to size_t now

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,276 @@
+
+
+To synchronize all the work-groups in a cluster, this extension proposes to 
+overload the `sycl::group_barrier` function, accepting the `cluster_group` type
```
Thanks for catching this, corrected the wording and also added a memory scope to go along with it

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,276 @@
+
+Because of the special scheduling guarantees associated with a cluster launch,
+the backend must know which kernel would be using this feature. This can be
+thought of as a runtime property of the kernel.This proposal defines a new
```
Removed the phrase - "can be thought of as a runtime property"

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+    class cluster_group {
+    public:
+    ...
+    static constexpr memory_scope fence_scope = 
```
It looks like `memory_scope::sycl_ext_oneapi_experimental_cuda_cluster_fence` is used but there's no definition, we should define a new member of the `memory_scope` enum class:

```
enum class memory_scope {
...
sycl_ext_oneapi_experimental_cuda_cluster_group,
...
};

inline constexpr auto sycl_ext_oneapi_experimental_cuda_memory_scope_cluster_group
  = memory_scope::sycl_ext_oneapi_experimental_cuda_cluster_group;
```

And then can we define the value of `fence_scope` in the `cluster_group` class to this enum value.

EwanC commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+
+|`cluster_group<Dimensions> nd_item::ext_oneapi_cuda_get_cluster_group()`
+|Returns the constituent `cluster_group` in the kernel, representing this
+`cluster_group`'s overall position in the `nd_range`
```
This table renders slightly strangely
![image](https://github.com/intel/llvm/assets/2334151/ea825517-4896-42b5-beff-8841b2241976)

Might want this as the column headers in bold, also the monospace isn't quite right.
```
|Method
|Description

```

This should also be added to a couple other tables in this doc.

EwanC commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+To obtain the total number of clusters in the kernel, and to obtain the 
+id of the cluster of the calling work-item, this extension proposes to add two
+new member functions the `nd_item` class, namely 
+`ext_oneapi_cuda_get_size` and `ext_oneapi_cuda_get_cluster_id`
```
```suggestion
`ext_oneapi_cuda_get_size` and `ext_oneapi_cuda_get_cluster_id`.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|===
+
+
+To obain the total number of clusters within the group, and getting the cluster id, `nd_item` to have 2 more class methods, namely `get_cluster_range` and `get_cluster_id`
```
This sounds good, though I would say `nd-range` rather than `kernel`.

AD2605 replied:
changed nd_range to kernel

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
```
Before this section we should have a "Contributors" section which lists everyone who has contributed to the extension.

AD2605 replied:
Added the contributors section

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+
+=== Backend support status 
+
+This extension is only supported by the `ext_oneapi_cuda` backend.
```
We should have this say the extension is not implemented, and then update this once the implementation PR has been merged, just to avoid this being incorrect due to PR merge order.

AD2605 replied:
Fixed

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+NVIDIA’s compute capability (cc) 9.0 devices introduced a new level in the
+thread hierarchy, called as thread block clusters, in CUDA terminology. A thread
+block cluster, is a collection of thread blocks (a work-group in SYCL
+terminology). The work-groups which make up a cluster have the ability to access
```
We should also mention here that an Nvidia thread block cluster represents a number of thread blocks which can execute concurrently.

AD2605 replied:
Added, 
It now says - 

> NVIDIA’s compute capability (cc) 9.0 devices introduced a new level in the
thread hierarchy, called as thread block clusters, in CUDA terminology. A thread
block cluster, is a collection of thread blocks (a work-group in SYCL
terminology) that run concurrently

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+each other's local memory, and can be synchronized. This has various
+applications, convolutions, GEMMs and FFTs to name a few.
+
+This proposal introduces a SYCL equivalent of the same, to be called as
```
```suggestion
This proposal introduces a SYCL API to expose this capability, to be called as
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+Because of the special scheduling guarantees associated with a cluster launch,
+the backend must know which kernel would be using this feature. Thus, this 
+proposal introduces a new launch property called as `cluster_size` that will
+contain the cluster  in terms of number of workgroups.
```
```suggestion
contain the cluster in terms of number of work-groups.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+struct cluster_size {
+  cluster_size(const sycl::range<Dim>& size) : 
+    size(size) {}
+  size_t operator[](int i) { return size[i]; }
```
```suggestion
  size_t operator[](int i)
```
We don't need a definition, just the declaration.

AD2605 replied:
Fixed, 
though I have changed this query, instead of returning the cluster size for a particular dimension, it now returns the sycl::range itself

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+|Description
+
+|`linear_id get_group_linear_id() const`
+|Returns the flattened id of the calling `group` within the cluster
```
I think we should say `linearized` instead of `flattened`.

AD2605 replied:
Fixed

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+|`id_type get_group_id() const`
+|Returns the id of the calling work-group along each dimension
+|===
+
```
I think we should also have the following member functions:

* `get_local_id`
* `get_local_range`
* `get_local_linear_range`
* `get_group_linear_range`

following the convention of [sycl_ext_oneapi_non_uniform_groups](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_non_uniform_groups.asciidoc).

AD2605 replied:
Similar to non uniform groups, should we also add a `leader` ?

AD2605 replied:
Addressed, and added a leader method as well, similar to every other group

AerialMantis replied:
Great thanks, yeah good catch, we should have a `leader` member function as well.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+[%header,cols="10,5"]
+|===
+
+|`range<3> nd_item::ext_oneapi_cuda_get_size(size_t) const`
```
I think we probably don't need member functions to return the size and id since we have `get_cluster_group`.

AD2605 replied:
`ext_oneapi_cuda_get_size` is now `ext_oneapi_cuda_get_cluster_range`, it somehow got named as ` ext_oneapi_cuda_get_size`. 

This is to get the total number of clusters in the kernel. similar to `nd_item::get_group_range`.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+all the work-groups within the cluster must exist and the work-group's should
+not cease to exist before all the memory operations are completed. This can be 
+ensured by synchronizing all the work-items within the cluster before and after
+the local memory operations.
```
```suggestion
the local memory operations, using `group_barrier`.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,326 @@
+|T*  map_cluster_local_pointer(T* addr, size_t group_id)
+|Accepts the equivalent address to the memory location relative to the calling 
+work-item which is to be mapped from the local memory of the work-group, as 
+specified by the group_id, denoting the flattened group_id within the cluster
```
```suggestion
specified by the group_id, denoting the linear group_id within the cluster
```

JackAKirk commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,379 @@
+contain the cluster in terms of number of work-groups.
+ 
+
+=== Extension to `enum class aspect`
```
From here to 
```
A device requires the `sycl_ext_oneapi_cuda_cluster_group` aspect to
support launching a kernel with the property defined in the following section.
```

needs to be moved before the section "Launching a kernel with a `cluster_group`"

At the moment the aspect description is stuck in the middle of "Launching a kernel with a `cluster_group`"

AD2605 replied:
Done

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc:
```diff
@@ -0,0 +1,200 @@
+|Returns the flattened index of the calling `workitem` within the cluster
+
+|`size_t get_group_linear_range(void)`
+|Returns the toal number of `groups` within the cluster
```
changed it back it get_group_range()

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+[source, c++]
+----
+size_t max_cluster_size = 
+      device.get_info<info::device::ext_cuda_experimental_max_cluster_group_size>();
```
Should this be:
```suggestion
      device.get_info<info::device::ext_oneapi_cuda_experimental_max_cluster_group_size>();
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+=== Accessing the Cluster Group From Device Code
+
+Building upon the group hierarchy in sycl, this proposal adds another level
```
```suggestion
Building upon the group hierarchy in SYCL, this proposal adds another level
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+    public:
+    ...
+    static constexpr memory_scope fence_scope = 
+                memory_scope::sycl_ext_oneapi_experimental_cuda_cluster_fence;
```
```suggestion
                memory_scope::sycl_ext_oneapi_experimental_cuda_cluster_group;
```

Also there seems to be some confusion about the ordering of the namespaces here.  Sometimes you have these expanded as "oneapi_experimental_cuda" and sometimes they're expanded as "oneapi_cuda_experimental".  Which is right?

AerialMantis replied:
So I'm not entirely sure which ordering is correct here, I've seen both used before, my instinct is that `oneapi_cuda_experimental` makes more sense, @gmlueck do we have a convention for one or the other?

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+    }
+----
+
+To synchronize all workitems in the cluster group, `sycl::group_barrier` can be 
```
```suggestion
To coordinate all work-items in the cluster group, `sycl::group_barrier` can be 
```

AerialMantis replied:
Fixed in https://github.com/intel/llvm/pull/13594/commits/bf131725ad90260b2496948fc5b1d6ab0c80d372.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+namespace sycl {
+enum class aspect {
+  ...
+  sycl_ext_oneapi_cuda_cluster_group
```
```suggestion
  ext_oneapi_cuda_cluster_group
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+}
+----
+
+A device requires the `sycl_ext_oneapi_cuda_cluster_group` aspect to
```
```suggestion
A device requires the `ext_oneapi_cuda_cluster_group` aspect to
```

AD2605 replied:
Fixed in [beabf](https://github.com/intel/llvm/pull/13594/commits/beabfadeeb7161fb5c5cbc8c69e9cfb9ad5bc888)

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+  cluster_size(const sycl::range<Dim>& size) : 
+    size(size) {}
+  sycl::range<Dim> get_cluster_size();
+  sycl::range<Dim> size;
```
We only need to specify the interface so we don't need to define the member var.

AD2605 replied:
Fixed in [beabf](https://github.com/intel/llvm/pull/13594/commits/beabfadeeb7161fb5c5cbc8c69e9cfb9ad5bc888)

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+[source, c++]
+----
+size_t max_cluster_size = 
```
We should add a definition for the info descriptor.

AD2605 replied:
Added in [beabf](https://github.com/intel/llvm/pull/13594/commits/beabfadeeb7161fb5c5cbc8c69e9cfb9ad5bc888)

Though @AerialMantis , should this be in the info::device namespace as well ?

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+Building upon the group hierarchy in SYCL, this proposal adds another level
+above `group`(for work-groups), to be called as `cluster-group`, which 
+represents a collection of work-groups and will be accessible via the `nd_item`
+class, via a method to be introduced called `ext_oneapi_cuda_get_cluster_group()`.
```
```suggestion
class, via a member function to be introduced called `ext_oneapi_cuda_get_cluster_group()`.
```

AD2605 replied:
Fixed in [beabf](https://github.com/intel/llvm/pull/13594/commits/beabfadeeb7161fb5c5cbc8c69e9cfb9ad5bc888)

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+work-item which is to be mapped from the local memory of the work-group, as 
+specified by the group_id, denoting the linear group_id within the cluster
+|===
+
```
We should also add a function in this extension which maps to `query_shared_rank`.

AD2605 replied:
Added in [beabf](https://github.com/intel/llvm/pull/13594/commits/beabfadeeb7161fb5c5cbc8c69e9cfb9ad5bc888)

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+----
+  enum class memory_scope {
+  ...
+  sycl_ext_oneapi_experimental_cuda_cluster_group,
```
```suggestion
  ext_oneapi_cuda_cluster_group,
```

AerialMantis replied:
This has been fixed.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+Work-items in a work-group can atomically operate on the local memory addresses
+of another work-group which have mapped as described above. To facilitate this,
+this proposal introduces a new memory scope, 
+`sycl_ext_oneapi_experimental_cuda_cluster_fence` which can be used with the
```
```suggestion
`ext_oneapi_cuda_cluster_fence` which can be used with the
```

AD2605 replied:
Fixed

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+[source,c++]
+----
+    class cluster_group {
```
We should move this to the earlier synopsis.

AD2605 replied:
Sorry, I did not get where to this exactly, could you please highlight it ?

AerialMantis replied:
I think this is fine now.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+|Method
+|Description
+
+|T*  map_cluster_local_pointer(T* addr, size_t group_id)
```
We've started moving towards a convention of having free functions rather than member functions, so I think we should make this a free function which takes a `cluster_group` as a parameter.

AerialMantis replied:
We should also add an overload which takes a `local_ptr`.

AerialMantis replied:
I don't think we've specified a convention for this, but in recent features or extensions we have started opting for free functions over member functions, I'm thinking the the group functions/algorithms and enqueue functions. So I was suggesting we continue this pattern.

Pennycook replied:
I'm not sure if this should be a member or a free function yet, but I want to point out that we should be careful.

Currently, every function in SYCL with a signature like `sycl::foo(Group g, ...)` is a cooperative operation that implies the function calls synchronize with one another.  Mapping a pointer isn't a cooperative operation, but a signature like `sycl::map_cluster_local_pointer(Group cluster, T* addr, size_t group_id)` would strongly imply that it was. If we did opt for a free function, I think we'd need to at least move the group argument, something like `sycl::map_cluster_local_pointer(T* addr, Group cluster, size_t group_id)`.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+= sycl_ext_oneapi_cuda_cluster_group
+
+:source-highlighter: coderay 
```
Comment.

AerialMantis replied:
Comment.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+of another work-group which have mapped as described above. To facilitate this,
+this proposal introduces a new memory scope, 
+`sycl_ext_oneapi_experimental_cuda_cluster_fence` which can be used with the
+existing `atomic_ref` class.
```
The `memory_scope` enumeration can be used with other APIs too.  I assume `memory_scope::ext_oneapi_cuda_cluster_group` can be used with any of them?  If so, it's a bit confusing to list only `atomic_ref` here.  For example, maybe say something like:

> Work-items in a work-group can atomically operate on and synchronize with local memory addresses of another work-group which have been mapped as described above. To facilitate this, this proposal introduces a new memory scope which indicates a memory ordering constraint that applies to all work-items in the same cluster.

If you feel you need more clarity, you could add another sentence at the end:

> This memory scope can be used with `atomic_ref` and other SYCL APIs that use `memory_scope`.

AerialMantis replied:
Thanks for the suggestion, I've updated it to use this wording, see https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+----
+
+A device requires the `sycl_ext_oneapi_cuda_cluster_group` aspect to
+support launching a kernel with the property defined in the following section.
```
```suggestion
support launching a kernel with the `cluster_size` property defined in the following section.
```

AerialMantis replied:
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+Because of the special scheduling guarantees associated with a cluster launch,
+the backend must know which kernel would be using this feature. Thus, this 
+proposal introduces a new launch property called as `cluster_size` that will
+contain the cluster in terms of number of work-groups.
```
```suggestion
contain the cluster size in described in a number of work-groups.
```

AerialMantis replied:
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+the cluster range in each dimension.
+
+The launch functions introduced in `sycl_ext_oneapi_enqueue_functions` can then
+be used to launch the kernel with the mentioned property.
```
```suggestion
be used to launch the kernel with the `cluster_size` property.
```

AerialMantis replied:
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+|`cluster_group<Dim> nd_item::ext_oneapi_cuda_get_cluster_group()`
+|Returns the constituent `cluster_group` in the kernel, representing this
+`cluster_group` 's overall position in the `nd_range`
```
```suggestion
`cluster_group`'s overall position in the `nd_range`
```

AerialMantis replied:
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+== Cluster Memory Fence Scope and Barrier
+
+Work-items in a work-group can atomically operate on the local memory addresses
+of another work-group which have mapped as described above. To facilitate this,
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+Therefore, to access another work-group's local memory, the address needs to be
+mapped such that the address in another work-group is addressable within the 
+calling work-item. Further, to access another another work-group's local memory,
+all the work-groups within the cluster must exist and the work-group's should
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+local memory. Typically addresses which reside in the local memory of a 
+work-group can only be addressed by the work-items of that work-group. 
+Therefore, to access another work-group's local memory, the address needs to be
+mapped such that the address in another work-group is addressable within the 
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+Work-group's within the cluster have the ability to access another work-group's 
+local memory. Typically addresses which reside in the local memory of a 
+work-group can only be addressed by the work-items of that work-group. 
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+== Accessing another work-group's local memory
+
+Work-group's within the cluster have the ability to access another work-group's 
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+|Returns the total number of clusters across each dimension.
+
+|`id<Dim> nd_item::get_cluster_id() const`
+|Returns the id of the cluster along each dimension.
```
I agree, I've removed these, see https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+|Description
+
+|`linear_id get_group_linear_id() const`
+|Returns the linearized id of the calling `group` within the cluster
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+----
+
+Note that the total number of work-groups in the kernel must be a multiple of
+the cluster range in each dimension.
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+defined in this specification.*
+
+=== Backend support status 
+This extension has not been implemented.
```
Makes sense, I've removed this, see https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+== Status
+
+This is an experimental extension specification for the `ext_oneapi_cuda`
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+sycl::event launch_kernel_with_cluster() {
+    using namespace sycl::ext::oneapi::experimental;
+    using namespace sycl::ext::oneapi::experimental::cuda;
```
Fixed in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+  };
+
+  inline constexpr auto sycl_ext_oneapi_experimental_cuda_memory_scope_cluster_group
+  = memory_scope::sycl_ext_oneapi_experimental_cuda_cluster_group;
```
Good catch, I've fixed this in https://github.com/intel/llvm/pull/13594/commits/3c722c023be09c2cce817a2aaab65f01f01c45c2.

Though I was wondering what our convention is for using "experimental" is this only for namespaces or do we also use it in member names as well? For example should this be `ext_oneapi_cuda_cluster_group` or `ext_oneapi_experimental_cuda_cluster_group`?

gmlueck replied:
Only for the namespace.  We decided it was too wordy when used in function prefixes.

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+[source, c++]
+----
+size_t max_cluster_size = 
+      device.get_info<info::device::ext_oneapi_cuda_experimental_max_cluster_group_size>();
```
Fixed in https://github.com/intel/llvm/commit/3c722c023be09c2cce817a2aaab65f01f01c45c2.

rolandschulz commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,394 @@
+|Description
+
+|T*  map_cluster_local_pointer(T* addr, size_t group_id)
+|Accepts the equivalent address to the memory location relative to the calling 
```
It doesn't say anything about address space. I assume this means general SYCL rules apply both in/out pointer are generic address space? And therefore this will map to `__nvvm_mapa`?
Cutlass calls `mapa.shared::cluster` (https://github.com/NVIDIA/cutlass/blob/56b46e2d13875b46b8f6a03f9f5ac91e2bfdc01a/include/cute/arch/cluster_sm90.hpp#L169). Would it be recommended to directly call `__nvvm_mapa_shared_cluster` instead of `map_cluster_local_pointer` for best performance. 
Why do we have this function at all and don't just let users directly call the built-in. Seems like this function is anyhow a very low level function.

AD2605 replied:
Hi Roland, thanks for your feedback.

You raise a valid point, and this  should be differentiating interfaces accepting generic address space and shared address space (read decorated), and accepting multi_ptr would be a good way to do that.

Yes, this function will practically be  just a wrapper around `__nvvm_mapa*`, as it's interface was designed in a bottom up fashion so to speak, by checking what PTX requires and thus what we need to pass from the top level.

But I would still advocate against relying (or rather suggesting ) on the users to inline asm or use compiler intrinsics, as it would not be a very SYCL way of doing things, and would add lead to complexity in the user's code.

rolandschulz replied:
I agree it is nicer to have a proper SYCL interface. But for these kind of low level functions the advantage is small and therefore we should only do it if it doesn't come at a performance impact. In case you plan to add the version with shared address space in a later version, I would prefer if you wait to add the whole function in that later version. Because I think it is better to recommend to use users inline asm or intrinsics over incurring a performance impact.

AerialMantis replied:
This raises a really good point, I've been discussing this with @Naghasan and we came up with some ideas for how to address this.

For the current interface, where we are simply using raw pointers, these would be in the generic address space, so this would map to a `mapa` instruction. We think this could be optimised to `mapa.shared::cluster` either by the LLVM NVPTX backend or perhaps in the PTX assembler, this would require some investigation, but without these optimizations we agree this would not be as efficient as calling the PTX instruction directly in asm. This would keep things simple, though it would be hiding the fact that a generic pointer may be to local memory and that may point to different local address spaces, which are CUDA-specific.

As @AD2605 mentioned we have considered having a `multi_ptr` interface as well and this may be a way to represent this. We could add an additional address space, though this would be a big change to introduce for this feature. The other thing to consider here is that LLVM currently only has one local memory address space so it cannot distinguish between `.shared::cta` and `.shared::cluster`, so again it would rely on the NVPTX backend to lower this correctly, perhaps inferring the correct address space through the explicit use of `mapa.shared::cluster`.

I think we tend to agree with @rolandschulz's conclusion that perhaps we are trying to expose a CUDA API which is too low level to abstract in SYCL without expanding the SYCL memory model. Another option we have considered, which we wanted to explore more in the future was whether it would be better to rather than exposing the low-level API for mapping local memory between work-groups in a cluster-group, we instead provide group algorithms for the `cluster_group` for all the existing algorithms. The mapping of local memory between work-groups in a cluster-group would then become an implementation detail of these algorithms. I think this would also address @Pennycook's concern above about having functions which appear like group functions but don't follow the same semantics, i.e. are not cooperative.

Our initial goal was to try to align with CUDA here, but perhaps in this case it's better to have a more high-level abstraction. We could still have a SYCLcompat interface which provides a more low-level mapping of the CUDA APIs, though there's always a concern there that this would miss out on some potential optimisation opportunities.

We would like to close this PR soon so we can have this in for the 2025.0 release, so I would propose that we keep the current interface, clarify that the pointers will be assumed to be in the generic address space, and add a known issue that we'd like to continue to explore the design space for this. Alternatively we could remove these APIs in the current version entirely if there are concerns with having it in it's current form.

AD2605 replied:
@rolandschulz ,  I have added this in the known issues section here - [824af1e](https://github.com/intel/llvm/pull/13594/commits/824af1e30203f42612a8c13bbe4740c91aeb1a2f)

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,404 @@
+
+This proposal introduces a SYCL API to expose this capability, to be called as
+`cluster_group`, and adds the methods to launch a kernel with a cluster range
+and accessing the various id's associated with the cluster
```
Fixed in [beabf](https://github.com/intel/llvm/pull/13594/commits/beabfadeeb7161fb5c5cbc8c69e9cfb9ad5bc888), though  this section has changed considerably

gmlueck commented:
It looks like there are still some "oneapi" that should be changed to "codeplay".  Can you search the document for "oneapi" and reexamine each?

AD2605 commented:
@gmlueck thank you for pointing that out, 
It should all be fixed now

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,421 @@
+facilitate this, a new memory scope is introduced to the `memory_scope` class;
+`ext_codeplay_cuda_cluster_group` which indicates a memory ordering
+constraint that applies to all work-items in the same cluster-group. This memory
+scope can be used with `atomic_ref`` and other SYCL APIs that use
```
```suggestion
scope can be used with `atomic_ref` and other SYCL APIs that use
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,421 @@
+
+namespace ext::codeplay::experimental::cuda {
+inline constexpr auto memory_scope_cluster_group = memory_scope::ext_codeplay_cuda_cluster_group;
+} // namespace ext::codeplay::experimental::cuda
```
What are these lines?  It seems like they are duplicated from the code snippet above.  Note that they are not in a source code block, so they are getting rendered as plain text.

AD2605 replied:
Thanks
Removed in [f544043](https://github.com/intel/llvm/pull/13594/commits/f544043f994a2edcdf777c9a95d9eac159df6b94)

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,421 @@
+    inline constexpr auto memory_scope_cluster_group
+      = memory_scope::ext_codeplay_cuda_cluster_group;
+  } // namespace ext::codeplay::experimental::cuda
+
```
```suggestion
} // namespace sycl
```

It would be good to close off this scope so the braces are balanced.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,421 @@
+    public:
+    ...
+    static constexpr memory_scope fence_scope = 
+                memory_scope::ext_codeplay_cuda_cluster_fence;
```
Personally, I find it confusing to add this separate code listing that shows the `fence_scope` member type alias.  I'd add this in the main synopsis for `cluster_group` above, so that there is a single synopsis that shows all the public members.

AD2605 replied:
Done

gmlueck replied:
I don't see any change to the example.  Maybe you forgot to push a commit?

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,421 @@
+sycl::event launch_kernel_with_cluster() {
+  using cluster_size_prop =
+    sycl::ext::codeplay::experimental::cuda::cluster_size
+  using namespace sycl::ext::codeplay::experimental::cuda;
```
Our BKM for examples is to never use `using namespace` because it becomes unclear which identifiers are in which namespace.  Instead, we usually define namespace aliases, and then show fully qualified names.  For example, something like:

```
namespace syclcp = sycl::ext::codeplay::experimental;
namespace syclex = sycl::ext::oneapi::experimental;

sycl::event launch_kernel_with_cluster() {
  sycl::queue q;

  sycl::nd_range<3> ndRange({4096, 4096, 32}, {32, 32, 1});
  syclex::properties props(syclcp::cuda::cluster_size({4, 4, 1}));
  syclex::launch_config config(ndRange, props);

  return syclex::submit_with_event(q, [&](sycl::handler& cgh){
      syclex::nd_launch(cgh, config, [=](sycl::nd_item<3> it) {
        auto cg = it.ext_codeplay_cuda_get_cluster_group();
        auto cgId = cg.get_group_id();
        ...
      });
  })
}
```
```

AD2605 replied:
Done

gmlueck replied:
I don't see any change to the example.  Maybe you forgot to push a commit?

AD2605 replied:
It got stashed when pulling the committed suggested changes, and in my mind, I already had committed it, 
hence it went missing..

This should be fixed now in - [7bc66b2](https://github.com/intel/llvm/pull/13594/commits/7bc66b201b0c97d120683a3147e6da8886be9852)

Thanks

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,421 @@
+
+|`cluster_group<Dim> nd_item::ext_codeplay_cuda_get_cluster_group()`
+|Returns the constituent `cluster_group` in the kernel, representing this
+`cluster_group`'s overall position in the `nd_range`
```
```suggestion
`cluster_group` object's overall position in the `nd_range`
```

The text you had was not formatting correctly.  It's usually awkward to make a C++ type name possessive (or plural).  Instead, I usually find a way to rewrite the sentence to avoid this.  In addition, this proposed text fixes the formatting problem.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+work-group can only be accessed by the work-items of that work-group. 
+Therefore, to access another work-group's local memory, the address needs to be
+mapped such that the address in another work-group is accessible within the 
+calling work-item. Further, to access another another work-group's local memory,
```
```suggestion
calling work-item. Further, to access another work-group's local memory,
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+
+  enum class memory_scope {
+  ...
+  ext_codeplay_cuda_cluster_fence,
```
You have inconsistent naming for this enumerator.  For example, below it is called `ext_codeplay_cuda_cluster_group`:

```
inline constexpr auto memory_scope_cluster_group
      = memory_scope::ext_codeplay_cuda_cluster_group;
```

And there are some text references using the name `ext_codeplay_cuda_cluster_group` also.

I think `ext_codeplay_cuda_cluster_group` is a better name.  None of the other `memory_scope` enumerators have the word "fence" in their name.

AD2605 replied:
Good point, I have restored this now in [64a9384](https://github.com/intel/llvm/pull/13594/commits/64a9384c5b78e77ddd3e0785c7419bbea888a541)

gmlueck replied:
There is another occurrence of `ext_codeplay_cuda_cluster_fence` to change.

AD2605 replied:
Sorry about that, 
Fixed

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+specified by the group_id, denoting the linear group_id within the cluster
+|===
+
+Conversely, `get_linear_cluster_group_id_for_local_pointer` will return the 
```
Changed in [48d6888](https://github.com/intel/llvm/pull/13594/commits/48d68882bb7b9c701efc4aae2a02f0bfbe5be1b2)

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+properties  cluster_launch_property{cluster_size({1, 2, 1})};
+----
+
+Note that the total number of work-groups in the kernel must be a multiple of
```
We should use the new format for notes here:

```
[_Note:_ This is a note. _{endnote}_]
```

AD2605 replied:
done in [824af1e](https://github.com/intel/llvm/pull/13594/commits/824af1e30203f42612a8c13bbe4740c91aeb1a2f)

Though I believe it is not rendering correctly ?

gmlueck replied:
It looks right to me.  What looks wrong to you?

AD2605 replied:
This is how I see it - 
```
[Note: the total number of work-groups in the kernel must be a multiple of the cluster size in each dimension. — end note]
```

was this expected ? I had assumed it would be rendered like when one uses the `NOTE:` syntax
```
NOTE: this is the note
```

AerialMantis replied:
I think this is correct, there's an example of how this renders in [another ext](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc).

gmlueck replied:
That's the expected rendering.

AD2605 replied:
Oh I see
Thanks !

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|Description
+
+|`linear_id get_group_linear_id() const`
+|Returns the linearized id of the calling work-group within the cluster
```
```suggestion
|Returns the linearized id of the calling work-group within the cluster.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|Returns the linearized id of the calling work-group within the cluster
+
+|`linear_id get_local_linear_id() const`
+|Returns the linearized index of the calling work-item within the cluster
```
```suggestion
|Returns the linearized index of the calling work-item within the cluster.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|Returns the number of work-groups in each dimension within the cluster
+
+|`id_type get_group_id() const`
+|Returns the id of the calling work-group along each dimension within the cluster
```
```suggestion
|Returns the id of the calling work-group along each dimension within the cluster.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|Returns the id of the calling work-group along each dimension within the cluster
+
+|`id_type get_local_id() const`;
+|Returns the id of calling work-item along each dimension within the cluster
```
```suggestion
|Returns the id of calling work-item along each dimension within the cluster.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|size_t  get_linear_cluster_group_id_for_local_pointer(T* addr)
+|Accepts a mapped pointer to the local memory of another work-group and returns
+the linearized id of the work-group in the cluster from whom's local memory the
+mapped pointer was obtained
```
```suggestion
mapped pointer was obtained.
```

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|Description
+
+|size_t  get_linear_cluster_group_id_for_local_pointer(T* addr)
+|Accepts a mapped pointer to the local memory of another work-group and returns
```
We should specify what is returned in the case that the pointer has not been mapped, I assume since this pointer would be a regular local memory pointer, the work-group id would be that of the current work-item.

AD2605 replied:
Good point, 
I will change the wording to it accepting a valid shared memory address,
 and also add that any other address other than shared memory leads to UB

AD2605 replied:
Done in [824af1e](https://github.com/intel/llvm/pull/13594/commits/824af1e30203f42612a8c13bbe4740c91aeb1a2f)

AD2605 commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,413 @@
+
+== Known Issues
+
+This Specification does not mention the forward progress guarantees of the 
```
Done in [824af1e](https://github.com/intel/llvm/pull/13594/commits/824af1e30203f42612a8c13bbe4740c91aeb1a2f)

AerialMantis reviewed (approved):
Looks good!

AD2605 commented:
@intel/llvm-gatekeepers  This is ready to Merge. Thanks

AerialMantis commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,420 @@
+----
+
+
+To synchronize all workitems in the cluster group, `sycl::group_barrier` can be 
```
```suggestion
To coordinate all work-items in the cluster group, `sycl::group_barrier` can be 
```

Ruyk commented on sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc:
```diff
@@ -0,0 +1,420 @@
+
+[%hardbreaks] 
+
+Copyright (C) 2024-2024 Intel Corporation.  All rights reserved.
```
```suggestion
Copyright (C) 2024-2024 Codeplay Corporation.  All rights reserved.
```
If this is a Codeplay extension then should be Codeplay (C)

AD2605 replied:
Thanks for catching this.
Committing this verbatim

Ruyk reviewed (changes requested):
Minor comment about (C)

Ruyk commented:
LGTM
